include/mppic/parameters_handler.hpp

```diff
@@ -0,0 +1,173 @@
+// Copyright 2022 FastSense, Samsung Research
+
+#pragma once
+
+#include <vector>
+#include <functional>
+#include <type_traits>
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_lifecycle/lifecycle_node.hpp"
+#include "nav2_util/node_utils.hpp"
+
+namespace mppi
+{
+
+enum class ParameterType
+{
+  Dynamic,
+  Static
+};
+
+class ParametersHandler
+{
+public:
+  using get_param_func_t = void (const rclcpp::Parameter & param);
+  using post_callback_t = void ();
+  using pre_callback_t = void ();
+
+  ParametersHandler() = default;
+  ParametersHandler(const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent)
+  {
+    node_ = parent;
+    auto node = node_.lock();
+    logger_ = node->get_logger();
+  }
+
+  void start()
+  {
+    auto node = node_.lock();
+    on_set_param_handler_ = node->add_on_set_parameters_callback(
+      std::bind(
+        &ParametersHandler::dynamicParametersCallback, this, std::placeholders::_1));
+  }
+
+  rcl_interfaces::msg::SetParametersResult dynamicParametersCallback(
+    std::vector<rclcpp::Parameter> parameters)
+  {
+    rcl_interfaces::msg::SetParametersResult result;
+
+    std::lock_guard<std::mutex> lock(parameters_change_mutex_);
+
+    for (auto & pre_cb : pre_callbacks_) {
+      pre_cb();
+    }
+
+    for (auto & param : parameters) {
+      const std::string & param_name = param.get_name();
+
+      if (auto callback = get_param_callbacks_.find(param_name);
+        callback != get_param_callbacks_.end())
+      {
+        RCLCPP_INFO(logger_, "Parameter %s found", param_name.c_str());
+        callback->second(param);
+      } else {
+        RCLCPP_WARN(logger_, "Parameter %s not found", param_name.c_str());
+      }
+    }
+
+    for (auto & post_cb : post_callbacks_) {
+      post_cb();
+    }
+
+    result.successful = true;
+    return result;
+  }
+
+  std::mutex * getLock()
+  {
+    return &parameters_change_mutex_;
+  }
+
+
+  template<typename T>
+  void addPostCallback(T && callback)
+  {
+    post_callbacks_.push_back(std::move(callback));
+  }
+
+  template<typename T>
+  void addPreCallback(T && callback)
+  {
+    pre_callbacks_.push_back(std::move(callback));
+  }
+
+  auto getParamGetter(std::string ns)
+  {
+    return [this, ns = std::move(ns)](auto & setting, std::string name, auto default_value,
+             ParameterType param_type = ParameterType::Dynamic) {
+             getParam(
+               setting, ns == "" ? name : ns + "." + name, std::move(
+                 default_value), param_type);
+           };
+  }
+
+  template<typename SettingT, typename ParamT>
+  void getParam(
+    SettingT & setting, std::string name, ParamT default_value,
+    ParameterType param_type = ParameterType::Dynamic)
+  {
+    auto node = node_.lock();
+
+    nav2_util::declare_parameter_if_not_declared(
+      node, name, rclcpp::ParameterValue(default_value));
+
+    setStaticParam<ParamT>(node, setting, name);
```

Comment:
What about this is "Static"?

Reply:
(overall this name seems incorrect for what its doing)

Reply:
setting the parameter is happen inside, renamed ti setParameter